File: Minimal/tests/Brontide.Minimal.Host.Tests/ComponentProviderRestartOwnershipTests.fs
namespace Brontide.Minimal.Host.Tests

open System
open System.Diagnostics
open System.IO
open System.Text.Json
open System.Threading
open System.Threading.Tasks
open NUnit.Framework
open Brontide.Minimal.Experimental.ComponentManagement
open Brontide.Minimal.Host

type private RestartTemporaryOwnership() =
    let root = Path.Combine(Path.GetTempPath(), $"brontide-cbi54-{Guid.NewGuid():N}")
    member _.Root = root
    member _.Path = Path.Combine(root, "restart.owner")
    interface IDisposable with
        member _.Dispose() =
            let rec remove attempt =
                try if Directory.Exists root then Directory.Delete(root, true)
                with :? IOException when attempt < 249 ->
                    Threading.Thread.Sleep 20
                    remove (attempt + 1)
            remove 0

[<TestFixture>]
type ComponentProviderRestartOwnershipTests() =
    let multiple action = Assert.Multiple(Action action)
    let runIdentity = ProviderRestartAttemptRunId.create "restart-run.test.1"
    let occurrence = OccurrenceId.create "occ.def.test.cooling-provider.1"
    let staged = ProviderArtifactSetId.create (String('A', 64))

    let acquire path owner lease =
        DurableProviderRestartOwnership.Acquire(
            path, ProviderRestartOwnerId.create owner, ProviderRestartOwnershipLeaseId.create lease,
            runIdentity, occurrence, staged)

    /// Acquires after a holder process was killed. A process that has reported exit has not necessarily
    /// had its file handles released by the kernel yet, so the first attempt can still see the lock
    /// held and answer `restart-ownership-busy` with no snapshot. Retrying on exactly that code waits
    /// for the release without waiting a fixed time for it, and without weakening what the caller then
    /// asserts: any other code is returned immediately, so a lock that is never released still fails
    /// the test rather than passing late.
    let acquireAfterProcessLoss path owner lease =
        let rec attempt index =
            let result = acquire path owner lease
            if result.Code <> "restart-ownership-busy" || index = 249 then result
            else
                Thread.Sleep 20
                attempt (index + 1)
        attempt 0

    let providerPath () =
        match Environment.GetEnvironmentVariable "BRONTIDE_MINIMAL_PROVIDER" with
        | null | "" -> Assert.Ignore "BRONTIDE_MINIMAL_PROVIDER does not name a built provider endpoint."; ""
        | path when not (File.Exists path) -> Assert.Ignore "BRONTIDE_MINIMAL_PROVIDER does not name a built provider endpoint."; ""
        | path -> Path.GetFullPath path

    let probe verb path = task {
        let start = ProcessStartInfo(providerPath(), UseShellExecute = false, RedirectStandardInput = true, RedirectStandardOutput = true)
        start.ArgumentList.Add($"--{verb}-exclusive-file={path}")
        use child =
            match Process.Start start |> Option.ofObj with
            | Some value -> value
            | None -> failwith "The provider probe process did not start."
        do! child.WaitForExitAsync()
        return child.ExitCode
    }

    [<Test>]
    member _.``CBI54 C1 ownership is bound to one restart lineage``() =
        use temporary = new RestartTemporaryOwnership()
        let acquired = acquire temporary.Path "owner-a" "lease-a"
        (acquired.Ownership.Value :> IDisposable).Dispose()
        let before = File.ReadAllBytes(temporary.Path + ".state")
        let mismatch = DurableProviderRestartOwnership.Acquire(
            temporary.Path, ProviderRestartOwnerId.create "owner-a", ProviderRestartOwnershipLeaseId.create "lease-a",
            ProviderRestartAttemptRunId.create "restart-run.other", occurrence, staged)
        multiple (fun () ->
            Assert.That(acquired.Code, Is.EqualTo "restart-ownership-acquired")
            Assert.That(mismatch.Code, Is.EqualTo "restart-ownership-lineage-mismatch")
            Assert.That(File.ReadAllBytes(temporary.Path + ".state"), Is.EqualTo(box before)))

    [<Test; Category("CrossProcess")>]
    member _.``CBI54 C2 one operating system owner excludes other processes``() = task {
        use temporary = new RestartTemporaryOwnership()
        let acquired = acquire temporary.Path "owner-a" "lease-a"
        let! blocked = probe "probe" temporary.Path
        Assert.That(blocked, Is.EqualTo 74)
        (acquired.Ownership.Value :> IDisposable).Dispose()
        let! available = probe "probe" temporary.Path
        Assert.That(available, Is.Zero)
    }

    [<Test>]
    member _.``CBI54 C3 every acquisition advances an atomic durable fence``() =
        use temporary = new RestartTemporaryOwnership()
        let first = acquire temporary.Path "owner-a" "lease-a"
        (first.Ownership.Value :> IDisposable).Dispose()
        let before = File.ReadAllBytes(temporary.Path + ".state")
        Directory.CreateDirectory(temporary.Path + ".state.tmp") |> ignore
        let refused = acquire temporary.Path "owner-b" "lease-b"
        let afterRefused = File.ReadAllBytes(temporary.Path + ".state")
        Directory.Delete(temporary.Path + ".state.tmp")
        let second = acquire temporary.Path "owner-b" "lease-b"
        multiple (fun () ->
            Assert.That(refused.Code, Is.EqualTo "restart-ownership-write-failed")
            Assert.That(afterRefused, Is.EqualTo(box before))
            Assert.That(second.Snapshot.Value.Epoch, Is.EqualTo 2L))
        (second.Ownership.Value :> IDisposable).Dispose()

    [<Test>]
    member _.``CBI54 C4 only the current live lease matches the journal``() =
        use temporary = new RestartTemporaryOwnership()
        let journalPath = Path.Combine(temporary.Root, "restart.journal")
        let journal = DurableProviderRestartAttemptJournal.Establish(
            journalPath, runIdentity, occurrence, staged,
            ProviderRestartPolicy.create 2 (TimeSpan.FromMinutes 1.0)).Journal.Value
        let before = File.ReadAllBytes journalPath
        let ownership = (acquire temporary.Path "owner-a" "lease-a").Ownership.Value
        Assert.That(ownership.IsCurrentFor journal.Snapshot, Is.True)
        (ownership :> IDisposable).Dispose()
        multiple (fun () ->
            Assert.That(ownership.IsCurrentFor journal.Snapshot, Is.False)
            Assert.That(File.ReadAllBytes journalPath, Is.EqualTo(box before)))

    [<Test>]
    member _.``CBI54 C5 released and superseded leases are stale``() =
        use temporary = new RestartTemporaryOwnership()
        let first = acquire temporary.Path "owner-a" "lease-a"
        (first.Ownership.Value :> IDisposable).Dispose()
        let second = acquire temporary.Path "owner-a" "lease-a"
        multiple (fun () ->
            Assert.That(first.Ownership.Value.Snapshot.IsLive, Is.False)
            Assert.That(second.Snapshot.Value.Epoch, Is.EqualTo 2L)
            Assert.That(second.Snapshot.Value.IsLive, Is.True))
        (second.Ownership.Value :> IDisposable).Dispose()

    [<Test; Category("CrossProcess")>]
    member _.``CBI54 C6 process loss relinquishes exclusivity without erasing history``() = task {
        use temporary = new RestartTemporaryOwnership()
        let first = acquire temporary.Path "owner-a" "lease-a"
        (first.Ownership.Value :> IDisposable).Dispose()
        let start = ProcessStartInfo(providerPath(), UseShellExecute = false, RedirectStandardInput = true, RedirectStandardOutput = true)
        start.ArgumentList.Add($"--hold-exclusive-file={temporary.Path}")
        use holder =
            match Process.Start start |> Option.ofObj with
            | Some value -> value
            | None -> failwith "The provider holder process did not start."
        let! held = holder.StandardOutput.ReadLineAsync()
        Assert.That(held, Is.EqualTo "held")
        // Strict while the holder is alive: exclusivity must be refused at the first attempt.
        Assert.That((acquire temporary.Path "owner-b" "lease-b").Code, Is.EqualTo "restart-ownership-busy")
        holder.Kill(true)
        do! holder.WaitForExitAsync()
        let recovered = acquireAfterProcessLoss temporary.Path "owner-b" "lease-b"
        multiple (fun () ->
            Assert.That(recovered.Code, Is.EqualTo "restart-ownership-acquired")
            Assert.That(recovered.Snapshot.Value.Epoch, Is.EqualTo 2L))
        (recovered.Ownership.Value :> IDisposable).Dispose()
    }

    [<Test>]
    member _.``CBI54 C7 inspection is bounded and fails closed``() =
        use temporary = new RestartTemporaryOwnership()
        Assert.That(DurableProviderRestartOwnership.Inspect(temporary.Path, runIdentity, occurrence, staged).Code,
            Is.EqualTo "restart-ownership-missing")
        let acquired = acquire temporary.Path "owner-a" "lease-a"
        (acquired.Ownership.Value :> IDisposable).Dispose()
        let bytes = File.ReadAllBytes(temporary.Path + ".state")
        bytes[0] <- bytes[0] ^^^ 0x7Fuy
        File.WriteAllBytes(temporary.Path + ".state", bytes)
        Assert.That(DurableProviderRestartOwnership.Inspect(temporary.Path, runIdentity, occurrence, staged).Code,
            Is.EqualTo "restart-ownership-corrupt")

    [<Test>]
    member _.``CBI54 C8 minimal executes the shared ownership model``() =
        use fixture = JsonDocument.Parse(File.ReadAllText(Path.Combine(
            TestContext.CurrentContext.TestDirectory, "component-management", "fixtures", "cbi54-restart-ownership-vectors.json")))
        let text (value: JsonElement) = value.GetString() |> Option.ofObj |> Option.defaultValue ""
        for vector in fixture.RootElement.GetProperty("vectors").EnumerateArray() do
            use temporary = new RestartTemporaryOwnership()
            let mutable ownership: DurableProviderRestartOwnership option = None
            let mutable snapshot: ProviderRestartOwnershipSnapshot option = None
            let mutable code = "none"
            for actionElement in vector.GetProperty("actions").EnumerateArray() do
                let action = text actionElement
                if action.StartsWith("acquire:", StringComparison.Ordinal) then
                    let parts = action.Split ':'
                    let result = acquire temporary.Path parts[1] parts[2]
                    code <- result.Code
                    ownership <- result.Ownership
                    snapshot <- result.Snapshot
                elif action = "release" then
                    (ownership.Value :> IDisposable).Dispose()
                    snapshot <- Some ownership.Value.Snapshot
                elif action = "inspect" then
                    let result = DurableProviderRestartOwnership.Inspect(temporary.Path, runIdentity, occurrence, staged)
                    code <- result.Code
                    snapshot <- result.Snapshot
            multiple (fun () ->
                Assert.That(code, Is.EqualTo(vector.GetProperty("expectedCode") |> text))
                Assert.That(snapshot.Value.Epoch, Is.EqualTo(vector.GetProperty("expectedEpoch").GetInt64()))
                Assert.That(snapshot.Value.Owner.Value, Is.EqualTo(vector.GetProperty("expectedOwner") |> text))
                Assert.That(snapshot.Value.Lease.Value, Is.EqualTo(vector.GetProperty("expectedLease") |> text))
                Assert.That(snapshot.Value.IsLive, Is.EqualTo(vector.GetProperty("expectedLive").GetBoolean())))
            ownership |> Option.iter (fun value -> (value :> IDisposable).Dispose())

File: Reference/tests/Brontide.Reference.Studio.Tests/ComponentProviderRestartOwnershipTests.cs
using System.Diagnostics;
using System.Text.Json;
using Brontide.Reference.Experimental.ComponentManagement;

namespace Brontide.Reference.Studio.Tests;

[TestFixture]
public sealed class ComponentProviderRestartOwnershipTests
{
    private static readonly ProviderRestartAttemptRunId RunId = ProviderRestartAttemptRunId.Create("restart-run.test.1");
    private static readonly OccurrenceId Occurrence = OccurrenceId.Create("occ.def.test.cooling-provider.1");
    private static readonly ProviderArtifactSetId Staged = ProviderArtifactSetId.Create(new string('A', 64));
    private static readonly ProviderRestartOwnerId Owner = ProviderRestartOwnerId.Create("owner-a");
    private static readonly ProviderRestartOwnershipLeaseId Lease = ProviderRestartOwnershipLeaseId.Create("lease-a");

    private sealed class TemporaryOwnership : IDisposable
    {
        public string Root { get; } = System.IO.Path.Combine(System.IO.Path.GetTempPath(), $"brontide-cbi54-{Guid.NewGuid():N}");
        /// <summary>
        /// Ownership is acquired for a journal, so the journal path is what the tests name. The lock
        /// and its state are derived from it rather than chosen, which is the property C2 now rests on.
        /// </summary>
        public string JournalPath => System.IO.Path.Combine(Root, "restart.journal");
        public string LockPath => JournalPath + ".ownership";
        public string StatePath => LockPath + ".state";
        public void Dispose()
        {
            for (var attempt = 0; attempt < 250; attempt++)
            {
                try
                {
                    if (Directory.Exists(Root)) Directory.Delete(Root, recursive: true);
                    return;
                }
                catch (IOException) when (attempt < 249)
                {
                    Thread.Sleep(20);
                }
            }
        }
    }

    private static ProviderRestartOwnershipAcquireResult Acquire(string path, string owner = "owner-a", string lease = "lease-a") =>
        DurableProviderRestartOwnership.Acquire(
            path, ProviderRestartOwnerId.Create(owner), ProviderRestartOwnershipLeaseId.Create(lease),
            RunId, Occurrence, Staged);

    /// <summary>
    /// Acquires after a holder process was killed. A process that has reported exit has not
    /// necessarily had its file handles released by the kernel yet, so the first attempt can still see
    /// the lock held and answer <c>restart-ownership-busy</c> with no snapshot. Retrying on exactly
    /// that code waits for the release without waiting a fixed time for it, and without weakening what
    /// the caller then asserts: any other code is returned immediately, so a lock that is never
    /// released still fails the test rather than passing late.
    /// </summary>
    private static ProviderRestartOwnershipAcquireResult AcquireAfterProcessLoss(
        string path, string owner, string lease)
    {
        for (var attempt = 0; ; attempt++)
        {
            var result = Acquire(path, owner, lease);
            if (result.Code != "restart-ownership-busy" || attempt == 249) return result;
            Thread.Sleep(20);
        }
    }

    private static async Task<int> ProbeAsync(string provider, string verb, string path, bool waitForHeld = false)
    {
        var start = new ProcessStartInfo(provider) { UseShellExecute = false, RedirectStandardInput = true, RedirectStandardOutput = true };
        start.ArgumentList.Add($"--{verb}-exclusive-file={path}");
        using var process = Process.Start(start)!;
        if (waitForHeld) Assert.That(await process.StandardOutput.ReadLineAsync(), Is.EqualTo("held"));
        if (waitForHeld)
        {
            process.Kill(entireProcessTree: true);
            await process.WaitForExitAsync();
        }
        else await process.WaitForExitAsync();
        return process.ExitCode;
    }

    private static string ProviderPath()
    {
        var path = Environment.GetEnvironmentVariable("BRONTIDE_REFERENCE_PROVIDER");
        if (path is null || !File.Exists(path)) Assert.Ignore("BRONTIDE_REFERENCE_PROVIDER does not name a built provider endpoint.");
        return Path.GetFullPath(path!);
    }

    [Test]
    public void Cbi54_C1_ownership_is_bound_to_one_restart_lineage()
    {
        using var temporary = new TemporaryOwnership();
        var acquired = Acquire(temporary.JournalPath);
        acquired.Ownership!.Dispose();
        var before = File.ReadAllBytes(temporary.StatePath);
        var mismatch = DurableProviderRestartOwnership.Acquire(
            temporary.JournalPath, Owner, Lease, ProviderRestartAttemptRunId.Create("restart-run.other"), Occurrence, Staged);
        Assert.Multiple(() =>
        {
            Assert.That(acquired.Code, Is.EqualTo("restart-ownership-acquired"));
            Assert.That(mismatch.Code, Is.EqualTo("restart-ownership-lineage-mismatch"));
            Assert.That(File.ReadAllBytes(temporary.StatePath), Is.EqualTo(before));
        });
    }

    [Test, Category("CrossProcess")]
    public async Task Cbi54_C2_one_operating_system_owner_excludes_other_processes()
    {
        using var temporary = new TemporaryOwnership();
        var acquired = Acquire(temporary.JournalPath);
        Assert.That(await ProbeAsync(ProviderPath(), "probe", temporary.LockPath), Is.EqualTo(74));
        acquired.Ownership!.Dispose();
        Assert.That(await ProbeAsync(ProviderPath(), "probe", temporary.LockPath), Is.Zero);
    }

    [Test]
    public void Cbi54_C3_every_acquisition_advances_an_atomic_durable_fence()
    {
        using var temporary = new TemporaryOwnership();
        var first = Acquire(temporary.JournalPath);
        first.Ownership!.Dispose();
        var before = File.ReadAllBytes(temporary.StatePath);
        Directory.CreateDirectory(temporary.StatePath + ".tmp");
        var refused = Acquire(temporary.JournalPath, "owner-b", "lease-b");
        var afterRefused = File.ReadAllBytes(temporary.StatePath);
        Directory.Delete(temporary.StatePath + ".tmp");
        var second = Acquire(temporary.JournalPath, "owner-b", "lease-b");
        Assert.Multiple(() =>
        {
            Assert.That(refused.Code, Is.EqualTo("restart-ownership-write-failed"));
            Assert.That(afterRefused, Is.EqualTo(before));
            Assert.That(second.Snapshot!.Epoch, Is.EqualTo(2));
        });
        second.Ownership!.Dispose();
    }

    [Test]
    public void Cbi54_C4_only_the_current_live_lease_matches_the_journal()
    {
        using var temporary = new TemporaryOwnership();
        var journal = DurableProviderRestartAttemptJournal.Establish(
            temporary.JournalPath, RunId, Occurrence, Staged,
            ProviderRestartPolicy.Create(2, TimeSpan.FromMinutes(1))).Journal!;
        var before = File.ReadAllBytes(temporary.JournalPath);
        var ownership = Acquire(temporary.JournalPath).Ownership!;
        Assert.That(ownership.IsCurrentFor(journal), Is.True);
        ownership.Dispose();
        Assert.Multiple(() =>
        {
            Assert.That(ownership.IsCurrentFor(journal), Is.False);
            Assert.That(File.ReadAllBytes(temporary.JournalPath), Is.EqualTo(before));
        });
    }

    /// <summary>
    /// The finding this capability exists for. Lineage identities do not identify a journal: a second
    /// journal carrying the same run, occurrence, and staged identity — which a copied journal file
    /// carries by construction — was previously indistinguishable from the one the lease was acquired
    /// for, so a lease could fence a journal nobody was excluded from. Deriving the lock path from the
    /// journal makes two hosts on one journal collide; comparing the path here catches a lease pointed
    /// at a different journal that the identity comparison alone accepts.
    /// </summary>
    [Test]
    public void Cbi54_C4_a_lease_fences_only_the_journal_it_names()
    {
        using var temporary = new TemporaryOwnership();
        var policy = ProviderRestartPolicy.Create(2, TimeSpan.FromMinutes(1));
        var owned = DurableProviderRestartAttemptJournal.Establish(
            temporary.JournalPath, RunId, Occurrence, Staged, policy).Journal!;
        var impostorPath = System.IO.Path.Combine(temporary.Root, "copied.journal");
        var impostor = DurableProviderRestartAttemptJournal.Establish(
            impostorPath, RunId, Occurrence, Staged, policy).Journal!;
        // Released even when an assertion below fails, so the lock never outlives the fixture and
        // wedges its own temporary-directory cleanup.
        using var ownership = Acquire(temporary.JournalPath).Ownership!;
        Assert.Multiple(() =>
        {
            // Identical lineage on both journals, so identity comparison alone cannot separate them.
            Assert.That(impostor.Snapshot.RunIdentity, Is.EqualTo(owned.Snapshot.RunIdentity));
            Assert.That(impostor.Snapshot.Occurrence, Is.EqualTo(owned.Snapshot.Occurrence));
            Assert.That(impostor.Snapshot.StagedIdentity, Is.EqualTo(owned.Snapshot.StagedIdentity));
            Assert.That(ownership.IsCurrentFor(owned), Is.True);
            Assert.That(ownership.IsCurrentFor(impostor), Is.False);
            Assert.That(ownership.JournalPath, Is.EqualTo(System.IO.Path.GetFullPath(temporary.JournalPath)));
        });
    }

    [Test]
    public void Cbi54_C5_released_and_superseded_leases_are_stale()
    {
        using var temporary = new TemporaryOwnership();
        var first = Acquire(temporary.JournalPath);
        first.Ownership!.Dispose();
        var second = Acquire(temporary.JournalPath);
        Assert.Multiple(() =>
        {
            Assert.That(first.Ownership.Snapshot.IsLive, Is.False);
            Assert.That(second.Snapshot!.Epoch, Is.EqualTo(2));
            Assert.That(second.Snapshot.IsLive, Is.True);
        });
        second.Ownership!.Dispose();
    }

    [Test, Category("CrossProcess")]
    public async Task Cbi54_C6_process_loss_relinquishes_exclusivity_without_erasing_history()
    {
        using var temporary = new TemporaryOwnership();
        var first = Acquire(temporary.JournalPath);
        first.Ownership!.Dispose();
        var start = new ProcessStartInfo(ProviderPath())
            { UseShellExecute = false, RedirectStandardInput = true, RedirectStandardOutput = true };
        start.ArgumentList.Add($"--hold-exclusive-file={temporary.LockPath}");
        using var holder = Process.Start(start)!;
        Assert.That(await holder.StandardOutput.ReadLineAsync(), Is.EqualTo("held"));
        // Strict while the holder is alive: exclusivity must be refused at the first attempt.
        Assert.That(Acquire(temporary.JournalPath, "owner-b", "lease-b").Code, Is.EqualTo("restart-ownership-busy"));
        holder.Kill(entireProcessTree: true);
        await holder.WaitForExitAsync();
        var recovered = AcquireAfterProcessLoss(temporary.JournalPath, "owner-b", "lease-b");
        Assert.Multiple(() =>
        {
            Assert.That(recovered.Code, Is.EqualTo("restart-ownership-acquired"));
            Assert.That(recovered.Snapshot!.Epoch, Is.EqualTo(2));
        });
        recovered.Ownership!.Dispose();
    }

    [Test]
    public void Cbi54_C7_inspection_is_bounded_and_fails_closed()
    {
        using var temporary = new TemporaryOwnership();
        Assert.That(DurableProviderRestartOwnership.Inspect(temporary.JournalPath, RunId, Occurrence, Staged).Code,
            Is.EqualTo("restart-ownership-missing"));
        var acquired = Acquire(temporary.JournalPath);
        acquired.Ownership!.Dispose();
        var bytes = File.ReadAllBytes(temporary.StatePath);
        bytes[0] ^= 0x7F;
        File.WriteAllBytes(temporary.StatePath, bytes);
        Assert.That(DurableProviderRestartOwnership.Inspect(temporary.JournalPath, RunId, Occurrence, Staged).Code,
            Is.EqualTo("restart-ownership-corrupt"));
    }

    [Test]
    public void Cbi54_C8_reference_executes_the_shared_ownership_model()
    {
        using var fixture = JsonDocument.Parse(File.ReadAllText(Path.Combine(
            TestContext.CurrentContext.TestDirectory, "component-management", "fixtures", "cbi54-restart-ownership-vectors.json")));
        foreach (var vector in fixture.RootElement.GetProperty("vectors").EnumerateArray())
        {
            using var temporary = new TemporaryOwnership();
            DurableProviderRestartOwnership? ownership = null;
            ProviderRestartOwnershipSnapshot? snapshot = null;
            var code = "none";
            foreach (var action in vector.GetProperty("actions").EnumerateArray().Select(value => value.GetString()!))
            {
                if (action.StartsWith("acquire:", StringComparison.Ordinal))
                {
                    var parts = action.Split(':');
                    var result = Acquire(temporary.JournalPath, parts[1], parts[2]);
                    code = result.Code; ownership = result.Ownership; snapshot = result.Snapshot;
                }
                else if (action == "release") { ownership!.Dispose(); snapshot = ownership.Snapshot; }
                else if (action == "inspect")
                {
                    var result = DurableProviderRestartOwnership.Inspect(temporary.JournalPath, RunId, Occurrence, Staged);
                    code = result.Code; snapshot = result.Snapshot;
                }
            }
            Assert.Multiple(() =>
            {
                Assert.That(code, Is.EqualTo(vector.GetProperty("expectedCode").GetString()));
                Assert.That(snapshot!.Epoch, Is.EqualTo(vector.GetProperty("expectedEpoch").GetInt64()));
                Assert.That(snapshot.Owner.Value, Is.EqualTo(vector.GetProperty("expectedOwner").GetString()));
                Assert.That(snapshot.Lease.Value, Is.EqualTo(vector.GetProperty("expectedLease").GetString()));
                Assert.That(snapshot.IsLive, Is.EqualTo(vector.GetProperty("expectedLive").GetBoolean()));
            });
            ownership?.Dispose();
        }
    }
}
